From 28cb74429d59c9e3d0c3333e4ee8f5f1dc7d6355 Mon Sep 17 00:00:00 2001
From: Jialin Zhang <zhangjialin11@huawei.com>
Date: Mon, 26 Dec 2022 23:13:48 +0800
Subject: [PATCH] timekeeping: Adding a padding before timekeeper in tk_core
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 28cb74429d59c9e3d0c3333e4ee8f5f1dc7d6355
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I47W8L
CVE: NA

---------------------------

Adding a padding before timekeeper in tk_core can futhor fix
the false sharing problem discovered by commit 605276666775
("timekeeping: Avoiding false sharing in field access of tk_core")

context1 of unixbench is tested on Huawei Kunpeng920.
before this patch
0 CPUs in system; running 96 parallel copies of tests

Pipe-based Context Switching               17213571.7 lps   (10.0 s, 1 samples)

System Benchmarks Partial Index              BASELINE       RESULT    INDEX
Pipe-based Context Switching                   4000.0   17213571.7  43033.9
                                                                   ========
System Benchmarks Index Score (Partial Only)                        43033.9

after this patch
0 CPUs in system; running 96 parallel copies of tests

Pipe-based Context Switching               20321268.3 lps   (10.0 s, 1 samples)

System Benchmarks Partial Index              BASELINE       RESULT    INDEX
Pipe-based Context Switching                   4000.0   20321268.3  50803.2
                                                                   ========
System Benchmarks Index Score (Partial Only)                        50803.2

Signed-off-by: Jialin Zhang <zhangjialin11@huawei.com>
Reviewed-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 kernel/time/timekeeping.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/kernel/time/timekeeping.c b/kernel/time/timekeeping.c
index e3142bf7648e..c8d9b89a7571 100644
--- a/kernel/time/timekeeping.c
+++ b/kernel/time/timekeeping.c
@@ -52,6 +52,9 @@ static struct {
 	u64		padding[8];
 #endif
 	seqcount_raw_spinlock_t	seq;
+#ifdef CONFIG_ARCH_LLC_128_LINE_SIZE
+	u64		padding2[2];
+#endif
 	struct timekeeper	timekeeper;
 #ifdef CONFIG_ARCH_LLC_128_LINE_SIZE
 } tk_core ____cacheline_aligned_128 = {
-- 
2.33.0

